File: android/edge-frameworks-core/src/main/kotlin/io/github/rajibsarwar/edgeframeworks/EdgeVectorMath.kt
package io.github.rajibsarwar.edgeframeworks

import kotlin.math.sqrt

object EdgeVectorMath {
    fun cosineSimilarity(
        left: EdgeEmbedding,
        right: EdgeEmbedding
    ): Float {
        if (left.values.size != right.values.size) {
            throw EdgeVectorException.DimensionMismatch(
                expected = left.values.size,
                actual = right.values.size
            )
        }

        var dot = 0f
        var leftMagnitude = 0f
        var rightMagnitude = 0f

        for (index in left.values.indices) {
            val l = left.values[index]
            val r = right.values[index]
            dot += l * r
            leftMagnitude += l * l
            rightMagnitude += r * r
        }

        if (leftMagnitude <= 0f || rightMagnitude <= 0f) {
            throw EdgeVectorException.ZeroMagnitude
        }

        return dot / (sqrt(leftMagnitude) * sqrt(rightMagnitude))
    }
}

File: ios/Sources/EdgeFrameworks/Runtime/EdgeVectorMath.swift
import Foundation

public enum EdgeVectorMath {
    public static func cosineSimilarity(
        _ lhs: EdgeEmbedding,
        _ rhs: EdgeEmbedding
    ) throws -> Float {
        guard lhs.values.count == rhs.values.count else {
            throw EdgeVectorError.dimensionMismatch(
                expected: lhs.values.count,
                actual: rhs.values.count
            )
        }

        var dot: Float = 0
        var lhsMagnitude: Float = 0
        var rhsMagnitude: Float = 0

        for (left, right) in zip(lhs.values, rhs.values) {
            dot += left * right
            lhsMagnitude += left * left
            rhsMagnitude += right * right
        }

        guard lhsMagnitude > 0, rhsMagnitude > 0 else {
            throw EdgeVectorError.zeroMagnitude
        }

        return dot / (sqrt(lhsMagnitude) * sqrt(rhsMagnitude))
    }
}
